File: src/main/java/at/searles/regexp/Concat.java
package at.searles.regexp;

class Concat implements Regexp {

    private final Regexp l;
    private final Regexp r;

    Concat(Regexp l, Regexp r) {
        this.l = l;
        this.r = r;
    }

    public Regexp l() {
        return l;
    }

    public Regexp r() {
        return r;
    }

    @Override
    public <A> A accept(Visitor<A> visitor) {
        return visitor.visitThen(l, r);
    }

    @Override
    public String toString() {
        return String.format("then(%s, %s)", l, r);
    }
}

File: src/main/kotlin/at/searles/parsing/lexer/regexp/Concat.kt
package at.searles.parsing.lexer.regexp

internal class Concat(private val l: Regexp, private val r: Regexp) : Regexp {
    fun l(): Regexp {
        return l
    }

    fun r(): Regexp {
        return r
    }

    override fun <A> accept(visitor: Visitor<A>): A {
        return visitor.visitThen(l, r)
    }

    override fun toString(): String {
        return String.format("then(%s, %s)", l, r)
    }

}